No static type error when destructuring "void"

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.4.0-dev.20190328

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
- destructure void

**Code**

```ts
function foo() {} // or function foo(): void {}
const {} = foo();
```

Needs tsconfig.json: 
```json
{ 
  "compilerOptions": { 
    "target": "es6",
    "strictNullChecks": true
  }
}
```

**Expected behavior:**
Expected `error TS2532: Object is possibly 'undefined'.`

You can get the error properly with this version:
```ts
function foo() { return undefined; }
const {} = foo();
```

**Actual behavior:**
No type error, but the resulting code crashes with the error `Cannot destructure 'undefined' or 'null'.`

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->
- [Expected error here](https://www.typescriptlang.org/play/index.html#src=function%20foo()%20%7B%0D%0A%20%20%20%20return%20undefined%3B%0D%0A%7D%0D%0A%0D%0Aconst%20%7B%7D%20%3D%20foo()%3B%0D%0A)
- [Compare to this](https://www.typescriptlang.org/play/index.html#src=function%20foo()%20%7B%0D%0A%20%20%20%20return%20undefined%3B%0D%0A%7D%0D%0A%0D%0Aconst%20%7B%7D%20%3D%20foo()%3B%0D%0A)

**Related Issues:** <!-- Did you find other bugs that looked similar? -->
- No

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -19518,6 +19518,21 @@ namespace ts {
             return type;
         }
 
+        function checkNonNullNonVoidType(
+            type: Type,
+            node: Node,
+            nullDiagnostic?: DiagnosticMessage,
+            undefinedDiagnostic?: DiagnosticMessage,
+            nullOrUndefinedDiagnostic?: DiagnosticMessage,
+            voidDiagnostic?: DiagnosticMessage
+        ): Type {
+            const nonNullType = checkNonNullType(type, node, nullDiagnostic, undefinedDiagnostic, nullOrUndefinedDiagnostic);
+            if (nonNullType !== errorType && nonNullType.flags & TypeFlags.Void) {
+                error(node, voidDiagnostic || Diagnostics.Object_is_possibly_undefined);
+            }
+            return nonNullType;
+        }
+
         function checkPropertyAccessExpression(node: PropertyAccessExpression) {
             return checkPropertyAccessExpressionOrQualifiedName(node, node.expression, node.name);
         }
@@ -26258,7 +26273,7 @@ namespace ts {
                 if (node.initializer && node.parent.parent.kind !== SyntaxKind.ForInStatement) {
                     const initializerType = checkExpressionCached(node.initializer);
                     if (strictNullChecks && node.name.elements.length === 0) {
-                        checkNonNullType(initializerType, node);
+                        checkNonNullNonVoidType(initializerType, node);
                     }
                     else {
                         checkTypeAssignableToAndOptionallyElaborate(initializerType, getWidenedTypeForVariableLikeDeclaration(node), node, node.initializer);

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -19518,17 +19518,10 @@ namespace ts {
             return type;
         }
 
-        function checkNonNullNonVoidType(
-            type: Type,
-            node: Node,
-            nullDiagnostic?: DiagnosticMessage,
-            undefinedDiagnostic?: DiagnosticMessage,
-            nullOrUndefinedDiagnostic?: DiagnosticMessage,
-            voidDiagnostic?: DiagnosticMessage
-        ): Type {
-            const nonNullType = checkNonNullType(type, node, nullDiagnostic, undefinedDiagnostic, nullOrUndefinedDiagnostic);
+        function checkNonNullNonVoidType(type: Type, node: Node): Type {
+            const nonNullType = checkNonNullType(type, node);
             if (nonNullType !== errorType && nonNullType.flags & TypeFlags.Void) {
-                error(node, voidDiagnostic || Diagnostics.Object_is_possibly_undefined);
+                error(node, Diagnostics.Object_is_possibly_undefined);
             }
             return nonNullType;
         }

--- a/tests/baselines/reference/destructuringVoid.js
+++ b/tests/baselines/reference/destructuringVoid.js
@@ -1,0 +1,7 @@
+//// [destructuringVoid.ts]
+declare const v: void;
+const {} = v;
+
+
+//// [destructuringVoid.js]
+var _a = v;

--- a/tests/baselines/reference/destructuringVoid.symbols
+++ b/tests/baselines/reference/destructuringVoid.symbols
@@ -1,0 +1,7 @@
+=== tests/cases/conformance/es6/destructuring/destructuringVoid.ts ===
+declare const v: void;
+>v : Symbol(v, Decl(destructuringVoid.ts, 0, 13))
+
+const {} = v;
+>v : Symbol(v, Decl(destructuringVoid.ts, 0, 13))
+

--- a/tests/baselines/reference/destructuringVoid.types
+++ b/tests/baselines/reference/destructuringVoid.types
@@ -1,0 +1,7 @@
+=== tests/cases/conformance/es6/destructuring/destructuringVoid.ts ===
+declare const v: void;
+>v : void
+
+const {} = v;
+>v : void
+

--- a/tests/baselines/reference/destructuringVoidStrictNullChecks.errors.txt
+++ b/tests/baselines/reference/destructuringVoidStrictNullChecks.errors.txt
@@ -1,0 +1,9 @@
+tests/cases/conformance/es6/destructuring/destructuringVoidStrictNullChecks.ts(2,7): error TS2532: Object is possibly 'undefined'.
+
+
+==== tests/cases/conformance/es6/destructuring/destructuringVoidStrictNullChecks.ts (1 errors) ====
+    declare const v: void;
+    const {} = v;
+          ~~
+!!! error TS2532: Object is possibly 'undefined'.
+    

--- a/tests/baselines/reference/destructuringVoidStrictNullChecks.js
+++ b/tests/baselines/reference/destructuringVoidStrictNullChecks.js
@@ -1,0 +1,7 @@
+//// [destructuringVoidStrictNullChecks.ts]
+declare const v: void;
+const {} = v;
+
+
+//// [destructuringVoidStrictNullChecks.js]
+var _a = v;

--- a/tests/baselines/reference/destructuringVoidStrictNullChecks.symbols
+++ b/tests/baselines/reference/destructuringVoidStrictNullChecks.symbols
@@ -1,0 +1,7 @@
+=== tests/cases/conformance/es6/destructuring/destructuringVoidStrictNullChecks.ts ===
+declare const v: void;
+>v : Symbol(v, Decl(destructuringVoidStrictNullChecks.ts, 0, 13))
+
+const {} = v;
+>v : Symbol(v, Decl(destructuringVoidStrictNullChecks.ts, 0, 13))
+

--- a/tests/baselines/reference/destructuringVoidStrictNullChecks.types
+++ b/tests/baselines/reference/destructuringVoidStrictNullChecks.types
@@ -1,0 +1,7 @@
+=== tests/cases/conformance/es6/destructuring/destructuringVoidStrictNullChecks.ts ===
+declare const v: void;
+>v : void
+
+const {} = v;
+>v : void
+

--- a/tests/cases/conformance/es6/destructuring/destructuringVoid.ts
+++ b/tests/cases/conformance/es6/destructuring/destructuringVoid.ts
@@ -1,0 +1,3 @@
+// @strictNullChecks: false
+declare const v: void;
+const {} = v;

--- a/tests/cases/conformance/es6/destructuring/destructuringVoidStrictNullChecks.ts
+++ b/tests/cases/conformance/es6/destructuring/destructuringVoidStrictNullChecks.ts
@@ -1,0 +1,3 @@
+// @strictNullChecks: true
+declare const v: void;
+const {} = v;

